### PR TITLE
[None][fix] Fix beam search log_probs non-determinism with batch_size > 1

### PR DESCRIPTION
**Pull request by achartier**
## Summary by CodeRabbit

* **Bug Fixes**
  * Corrected batch-slot handling in beam-search decoding so batched inference computes log-probabilities consistently across items and runs.
  * Ensured deterministic beam-search outputs for variable-length/unevenly finishing items.

* **Tests**
  * Added integration test and validator to assert log-prob determinism for Whisper beam-search across repeated runs.

## Description

`bh.nMaxBatchSize` in `BeamSearchLayer::forwardAsync` was derived from `outputIdsPtr->getDimension<0>()`, which reflects the *active* batch size. `outputIdsPtr` is sliced to `activeBatchSize` in `DynamicDecodeLayer::setupOutputIds` (`ITensor::slice(mOutputIdsPtrDevice, 0, batchSize)`), so as requests finish at different times and the batch drains unevenly, this dimension shrinks. BeamStage3 then reads `logProbsTiled` with the wrong stride, producing non-deterministic log_prob values across runs.

**Fix 1** (`beamSearchLayer.cu`): Use `outputIds->getDimension<0>()` unconditionally — `outputIds` is never sliced to active batch size (it is used as a base pointer with per-slot offsets up to `maxBatchSize-1`), so its dim-0 is always `maxBatchSize`. This supersedes a prior partial fix that only applied when `output_log_probs=True`.

**Fix 2** (`decodingKernels.cu` / `gptDecoderBatched.cpp`): `gatherTree`'s `insertUnfinishedPathKernel` was using `nMaxBatchSize=1` (from the per-slot `batchSize`) and reading `logProbsTiled` from the wrong base pointer (missing `batchSlot * beamWidth` offset). Fixed to use the actual tensor allocation dimension and correct pointer offset.

Both fixes are TRT backend only and affect all models using beam search with `batch_size > 1` and `output_log_probs=True` (or `gather_generation_logits=True`).

## Test Coverage

New integration test `test_whisper_log_probs_determinism` in `tests/integration/defs/examples/test_whisper.py`:
- Runs Whisper large-v3 with `batch_size=4`, `num_beams=4` across 5 inference runs
- Uses 4 different LibriSpeech samples so each batch item produces a different number of output tokens (uneven batch drain)
- Asserts `log_probs` are bit-identical across all runs
- Uses pre-cached local dataset (`llm_datasets_root`) — no HuggingFace download at test time

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tests/integration/defs/examples/validate_whisper_log_probs_determinism.py [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+                output_generation_logits=True,
+                output_log_probs=True,
+                return_dict=True,
+            )
+        torch.cuda.synchronize()
+        all_log_probs.append(outputs["log_probs"][:, 0, :].cpu())
+
+    max_diff = max(
+        (all_log_probs[0] - all_log_probs[i]).abs().max().item() for i in range(1, args.num_runs)
+    )
```
[the remark this answers is not in the document]
Comment (achartier) [reply tests/integration/defs/examples/validate_whisper_log_probs_determinism.py:176]: Fixed in last version, please check

File: tests/integration/defs/examples/test_whisper.py [hunk 2]
```diff
@@ -211,3 +211,84 @@ def test_whisper_beam_search_generation_logits(llm_venv, engine_dir,
...
+
+
+@skip_post_blackwell
+@pytest.mark.parametrize("num_beams", [4],
+                         ids=lambda num_beams: f'nb:{num_beams}')
+@pytest.mark.parametrize("batch_size", [4],
+                         ids=lambda batch_size: f'bs:{batch_size}')
+@pytest.mark.parametrize("whisper_model_root", ['large-v3'], indirect=True)
+def test_whisper_log_probs_determinism(llm_venv, engine_dir,
+                                       whisper_example_root, whisper_model_root,
```
Comment (stnie) [inline tests/integration/defs/examples/test_whisper.py:223]: Could you verify if this test is executed during CI?  In the test lists I only found explicit tests for `test_llm_whisper_general`.  See e.g. [test_list](https://github.com/NVIDIA/TensorRT-LLM/blob/main/tests/integration/test_lists/test-db/l0_a10.yml#L201)

File: cpp/tensorrt_llm/kernels/decodingKernels.cu [hunk 3]
```diff
@@ -781,15 +781,23 @@ void gatherTree(DecodingOutput const& decodingOutput, DecodingInput const& decod
...
+    bh.nMaxBatchSize = logProbsTiledMaxBatchSize;
     bh.nBatchSize = batchSize;
     bh.nBeamWidth = beamWidth;
     bh.nMaxSeqLen = maxSeqLength;
     bh.lengthPenalties = bufferCast<float>(*lengthPenaltyPtr);
     bh.inputLengths = bufferCast<SizeType32>(*decodingInput.lengths);
     bh.outputIds = bufferCast<TokenIdType>(finalOutputIds);
     bh.logProbs = bufferCastOrNull<float>(decodingOutput.logProbs);
-    bh.logProbsTiled = bufferCast<float>(*decodingOutput.logProbsTiled);
+    bh.logProbsTiled = bufferCast<float>(*decodingOutput.logProbsTiled) + batchSlot * beamWidth;
```
Comment (stnie) [inline cpp/tensorrt_llm/kernels/decodingKernels.cu:809]: Is it ensured that `logProbsTiled` does not go Out of bounds here? E.g. if `batchSlot > maxBatchSize` ? 
If not we should guardrail this.

File: cpp/tensorrt_llm/kernels/decodingKernels.cu [hunk 4]
```diff
@@ -781,15 +781,23 @@ void gatherTree(DecodingOutput const& decodingOutput, DecodingInput const& decod
...
-    bh.nMaxBatchSize = batchSize;
+    // logProbsTiled has shape [MSL, maxNumSequences, BM] and is passed unsliced.
+    // nMaxBatchSize must equal the allocation stride (dim-1), not the per-slot batchSize=1.
+    // The pointer is pre-offset by batchSlot*BM so that insertUnfinishedPathKernel,
+    // which uses bid=0 / nBatchSize=1, computes:
+    //   (base + batchSlot*BM)[step * maxBS * BM + 0*BM + beamIdx]
+    //   = base[step * maxBS * BM + batchSlot * BM + beamIdx]
+    //   = logProbsTiled[step][batchSlot][beamIdx]  ✓
+    auto const logProbsTiledMaxBatchSize = static_cast<SizeType32>(decodingOutput.logProbsTiled->getShape().d[1]);
+    bh.nMaxBatchSize = logProbsTiledMaxBatchSize;
```
Comment (stnie) [inline cpp/tensorrt_llm/kernels/decodingKernels.cu:801]: Would it work to just slice `logProbsTiled`, similar to the other buffers instead of passing it unsliced?  If I understand correctly this would resolve your other changes as well.

Comment (achartier) [reply cpp/tensorrt_llm/kernels/decodingKernels.cu:801]: `logProbsTiled` has batch size as dim 1 while the other buffers have it as dim 0. Since it's not contiguous memory, we can't slice it directly.

File: tests/integration/defs/examples/test_whisper.py (hunk 2 above)
Comment (achartier) [reply tests/integration/defs/examples/test_whisper.py:223]: Added in 615417ee4511ff666e40e4a72765bc4da5d04f02

File: cpp/tensorrt_llm/kernels/decodingKernels.cu (hunk 3 above)
Comment (achartier) [reply cpp/tensorrt_llm/kernels/decodingKernels.cu:809]: Added a check in 615417ee4511ff666e40e4a72765bc4da5d04f02

Comment (stnie) [reply cpp/tensorrt_llm/kernels/decodingKernels.cu:809]: For consistency I would propose to also check `beamWidth` or the product `beamWidth * batchSlot`.  To ensure that `bufferCast<float>(*decodingOutput.logProbsTiled) + batchSlot * beamWidth;` definitely points to a valid memory location.

File: tests/integration/defs/examples/validate_whisper_log_probs_determinism.py [hunk 5]
```diff
@@ -0,0 +1,184 @@
...
+def main():
+    args = parse_arguments()
+    tensorrt_llm.logger.set_level("warning")
+
+    engine_dir = Path(args.engine_dir)
+    encoder_config = read_config("encoder", engine_dir)
+    decoder_config = read_config("decoder", engine_dir)
+
+    n_mels = encoder_config["n_mels"]
+    is_multilingual = decoder_config["vocab_size"] >= 51865
```
Comment (stnie) [inline tests/integration/defs/examples/validate_whisper_log_probs_determinism.py]: Maybe add a short comment to explain the `51865`

File: cpp/tensorrt_llm/kernels/decodingKernels.cu [hunk 6]
```diff
@@ -781,15 +781,27 @@ void gatherTree(DecodingOutput const& decodingOutput, DecodingInput const& decod
...
     bh.nBatchSize = batchSize;
     bh.nBeamWidth = beamWidth;
     bh.nMaxSeqLen = maxSeqLength;
     bh.lengthPenalties = bufferCast<float>(*lengthPenaltyPtr);
     bh.inputLengths = bufferCast<SizeType32>(*decodingInput.lengths);
     bh.outputIds = bufferCast<TokenIdType>(finalOutputIds);
     bh.logProbs = bufferCastOrNull<float>(decodingOutput.logProbs);
-    bh.logProbsTiled = bufferCast<float>(*decodingOutput.logProbsTiled);
+    bh.logProbsTiled = bufferCast<float>(*decodingOutput.logProbsTiled) + batchSlot * beamWidth;
```
Comment (stnie) [inline cpp/tensorrt_llm/kernels/decodingKernels.cu:809]: IIUC. This change indicates that `logprobs` were previously not gathered from the correct `batchSlot` when `batchSlot > 0` .  Which should have caused wrong `logprobs.` 

Is there a test that checks that the `logprobs` are correct for all requests? 
If not such a test should be added. (i.e. check if the `logprobs` are in line with the `logits`).

**Review by stnie — COMMENTED** [review]
Added some comments regarding the testing.

File: cpp/tensorrt_llm/kernels/decodingKernels.cu (hunk 3 above)
Comment (achartier) [reply cpp/tensorrt_llm/kernels/decodingKernels.cu:809]: Added in 67214d4c1e9f08e5914995113b7de8874f8dc384

File: tests/integration/defs/examples/validate_whisper_log_probs_determinism.py (hunk 5 above)
Comment (achartier) [reply tests/integration/defs/examples/validate_whisper_log_probs_determinism.py]: Added in 67214d4c1e9f08e5914995113b7de8874f8dc384

File: cpp/tensorrt_llm/kernels/decodingKernels.cu (hunk 6 above)
Comment (achartier) [reply cpp/tensorrt_llm/kernels/decodingKernels.cu:809]: Expanded test in 67214d4c1e9f08e5914995113b7de8874f8dc384

**Review by stnie — APPROVED** [review]
